11_taxonomic_information.Rmd

```diff
@@ -164,8 +164,65 @@ assay(altExp(tse, "Family"), "counts")[1:5,1:7]
 ```
 
 `altExpNames` now consists of `Family` level data. This can be extended to use 
-any taxonomic level listed in `mia::taxonomyRanks(tse)`.   
+any taxonomic level listed in `mia::taxonomyRanks(tse)`.
 
+### Taxa clustering {#taxa-clustering}
+Another way to agglomerate the data is to do clustering on the taxa. To do so,
+we usually start by doing a compositionality aware transformation such as CLR, 
+followed by the application of a standard clustering method. 
+
+Here is an example that does a CLR transformation followed by the hierarchical
+clustering algorithm. 
+
+First, we import the library `bluster` that simplifies the clustering. We
+
+```{r bluster_dependence}
+library(bluster)
+```
+
+Then we do the CLR transform followed by the clustering.
+
+```{r taxa_clustering}
+# The result of the CLR transform is stored in the assay clr
+tse <- transformCounts(tse, assay.type = "counts", method = "clr", 
+                       pseudocount = 1, MARGIN = "features")
+
+# We do the clustering on the features of the clr assay 
+tse <- cluster(tse, assay.type = "clr", HclustParam(method = "ward.D2"), 
+               MARGIN = "features", clust.col = "hclust")
```

Comment:
So the distance here is a custom function? Because I didn't find anything on `dist` or `vegdist`. I should have function like this?
```r
dist <- function(x,y) {
    1 - cor.test(x,y, method="kendall")
}
```

Reply:
The `hclust` function takes distance matrix `d` as input. That is by default calculated with base `dist` I think. There the default distance is Euclidean. I think that the `HclustParam` calls these functions?

Reply:
By the way, the `dist` function uses parameter `method` to refer to distances, hclust uses `method` to refer to the clustering method.. this might be an issue for argument passing..

Reply:
Yes, HclustParam first computes the distance matrix by calling the function passed in `metric` param, with the matrix and the `method` parameter, then it calls `hclust`.

I was also a bit surprised for the parameter `method` and there was an error in bluster linked to that, but it's fixed now so it works as it should. The parameters are documented so there shouldn't be any confusion

Reply:
So what function do you want me to call for HclustParam? Should I let the default "euclidean", or do you want something else? Because I may have not understood what you meant by 

> dissimilarity 1 - tau

Reply:
It is still essential to show how the distance metrics is set. It is very likely that a user may want to experiment with other distance metrics (or dissimilarity methods).

Reply:
I would use Kendall's tau as the similarity measure.

But that has inverse sign (higher values indicate higher similarity). Therefore the dissimilarity measure is 1-tau.

This is not available in `dist` or `vegdist` I think. 

Can we provide a custom dissimilarity function for this? Then you could show how to do it for Euclid, and how to do it for 1-tau?

Reply:
we can provide a custom function yes, I'm just unsure on how to program that function.
Would this be ok?
```r
tse <- cluster(tse, assay.type = "clr", HclustParam(method = "ward.D2", dist.fun = function(x) {
      kendall_distance <- function(x, y) {
            cor_test <- cor.test(x, y, method = "kendall")
            1 - cor_test$estimate
      }
      for (i in 1:(n-1)) {
            for (j in (i+1):n) {
                  distance_matrix[i, j] <- kendall_distance(input_matrix[i, ], input_matrix[j, ])
                  distance_matrix[j, i] <- distance_matrix[i, j]
            }
      }
}), 
               MARGIN = "features", clust.col = "hclust")
```

Reply:
1. better to define the custom function outside the cluster command and just call the function in there
2. you could replace `cor.test` with `cor`, and this way you can calculate correlations between all column pairs without a for loop, just like `cor(x,y, method="kendall")` for matrices x and y. 
3. i don't think it is a distance, I guess it is dissimilarity